core/src/main/java/org/apache/druid/common/config/NullHandling.java

```diff
@@ -50,16 +50,22 @@
    * It does not take effect in all unit tests since we don't use Guice Injection.
    */
   @Inject
-  private static NullValueHandlingConfig INSTANCE = new NullValueHandlingConfig(
-      Boolean.valueOf(System.getProperty(NULL_HANDLING_CONFIG_STRING, "true"))
-  );
+  private static NullValueHandlingConfig INSTANCE;
+
+  private static volatile Boolean isReplaceWithDefault = null;
```

Comment:
I'm not sure having a volatile static is a good idea for something that will be used by multiple threads while processing every row. Your benchmark checked single-threaded performance but this could degrade under multithreading.

Could you please either look into this, or do a hybrid of your last two approaches (require explicit initialization in tests, of a non-volatile static, and if-guard it in methods that use it).

Reply:
After discussion, maybe the explicit initialization is better in order to rid magic from our world, closing this PR and re-opening #8876